### PR TITLE
Feature: Configure per-request timeout setting via env[SERVICE_TIMEOUT_KEY]

### PR DESCRIPTION
**Pull request by TSMMark**
First, thanks for the awesome middleware!

I need the ability to customize the timeout value on a per-request basis, determined by some external dynamic data. This seemed like the most straightforward way to accomplish that. 

Let me know if this is something that could get merged, and if so what kind of tests and documentation would be necessary.

FWIW I smoke tested this patch in my project and it works as intended.

### CODE REVIEW

File: lib/rack/timeout/core.rb [hunk 1]
```diff
@@ -44,6 +44,7 @@ def ms(k)   # helper method used for formatting values in milliseconds
       :completed, # We're done with this request (also set after having timed out a request)
       ]
     ENV_INFO_KEY = "rack-timeout.info".freeze # key under which each request's RequestDetails instance is stored in its env.
+    ENV_SERVICE_TIMEOUT_KEY = "rack-timeout.service_timeout".freeze # key under which each request's RequestDetails instance is stored in its env.
```
Comment (dentarg) [inline lib/rack/timeout/core.rb:47]: Comment looks incorrect to me

**Review by dentarg — CHANGES_REQUESTED** [review]
(Add a test? Update the docs?) Sorry, didn't read the full description above 🙄

**Comment by GBIDEV** [comment]
@TSMMark Can you show us how you have used this in your tests?

Show an example of what changes you made to the conifg and parameters to get this working.

**Comment by wuputah** [comment]
Hello and thank you for your interest and contribution. Unfortunately this is a feature that has been added by many different contributors but has been rejected many times as it goes against the core philosophy of the middleware. The general suggestion is any requests that need to run for more than your `service_timeout` need to be moved to a background system. You can easily impact the performance of the rest of your application by allowing some requests to run longer.

In addition, changing an environment variable on each request is also not a great approach as it is not safe in a concurrent environment, and environment variables are generally best left unchanged.

**Comment by TSMMark** [comment]
@wuputah Please review the code and not just the PR title. We're only modifying the Rack request `env`, which is not the same as global `ENV`, and AFAIK is per-request object and is safe to be modified at will to contain request-specific information.

I do not agree that this is against the philosophy of the middleware. As a developer you could already easily segment out different portions of your routes into different sinatra / cuba / roda instances, each with their own individual middleware (and individual timeout settings). This is just a different way to do that.

Please consider reopening the PR. I will be happy to add specs

**Comment by TSMMark** [comment]
Following the history of this topic, 

#160 was a bad implementation
#138 was a bad implementation 
#110 was a bad implementation
etc, etc 

> [So, in short, this lets people shoot themselves in the foot either a bit more or a bit less, but rack-timeout is all about foot-shooting anyway, so does it really matter?](https://github.com/sharpstone/rack-timeout/pull/110#issuecomment-243524073)

Great point, sounds like we should introduce this feature since tens of people are asking for it and it doesn't cost anything.

**Comment by wuputah** [comment]
Ah, sorry, I misread this line as `ENV.fetch` rather than `env.fetch`.

```ruby
 service_timeout_env = env.fetch(ENV_SERVICE_TIMEOUT_KEY, service_timeout)
```

So you are setting this in your app via `request.env[ENV_SERVICE_TIMEOUT_KEY]`  before the middleware fires? Are you using another middleware inserted in the stack to do this? It's certainly a minimal approach as far as impact to Rack Timeout, but maybe not very user friendly.

----

#110 was the last time this discussion happened in depth. Ultimately other PRs were not necessarily closed due to implementation or design, but ultimately it was a decision to not implement variable timeouts in Rack Timeout. After all the discussions that have happened, it's pretty ingrained in my mind not to merge a PR for this feature. 😂 😭 

Additionally, the feature does exist (at least via one implementation) in slowpoke for those that need it.

**Comment by TSMMark** [comment]
> Are you using another middleware inserted in the stack to do this

Yes

> It's certainly a minimal approach as far as impact to Rack Timeout, but maybe not very user friendly.

Yeah exactly, I didn't want to impact performance, footprint, or maintainability of this library, while giving users who really need this functionality an escape hatch to do so.

> slowpoke

Yeah slowpoke seems to be pretty Rails-specific, which I am not using.

---

So due to the limited impact on this library would you consider a merge or am I going to have to maintain my fork indefinitely?

**Comment by dentarg** [comment]
@wuputah Wouldn't it be resonable to offer something like this? It is even more minimal than #110.

I don't think the [arguments](https://github.com/sharpstone/rack-timeout/pull/110#issuecomment-243524073) in #110 hold up, you said it yourself at https://github.com/sharpstone/rack-timeout/issues/169#issuecomment-806077437

> Additionally, while the application may enter a bad state when using Rails + ActiveRecord, the middleware does not know what you are using -- your app might only be making network calls, for instance.

Isn't the above an argument for legitimate use-cases of rack-timeout? (not just using it as a debugging tool) Even if using it as a debugging tool, offer this flexibility could help IMHO.

**Comment by DJSAMPAT** [comment]
@TSMMark It might be best to add all the specs and steps and update the docs with steps to show how you used it and then, if possible, have a version with it and one without, so users can fork the one linked to this without having to maintain it.

**Comment by TSMMark** [comment]
Just pushed a commit that adds a test. Although the example is contrived and it's not probably not a good idea to use a query param for this, it was an easy way to add a dynamic element other than path. 

https://github.com/sharpstone/rack-timeout/commit/2dc21dd350016e035f01500db06b2ed1088b996b

While doing that I noticed that the error messages actually include a trailing space character, and one of them ends in a period while the other does not, which is kind of weird or unexpected. Would you like a PR to clean up those messages?

**Pull request closed without merging**